### ISSUE TITLE
When switching scene collections, the new scene collection often loads before the old one has fully unloaded, causing a variety of issues.

### ISSUE THREAD

**Issue by skeletonbow** [obsproject/obs-studio#2715]
## Platform

Operating system and version:  Windows 10 x64 (1909)
OBS Studio version:  25.0.4 (and all other versions for the last 2 years or so)

## Expected Behavior

When switching from one scene collection to another while streaming live and/or recording, any video capture devices which are used in both collections should continue to work when OBS
switches from one collection to another.

## Current Behavior

When starting in scene collection 1, the webcam works as expected.  After switching to another scene collection (any) which use the same webcam, the webcam will not work by default.  It can sometimes be coaxed into working by bring up the video capture properties for the webcam and disconnecting and reconnecting it, however sometimes this doesn't work and restarting OBS is required.

Please note that with a simple OBS configuration with just a scene with a webcam the problem probably wont happen.  It becomes increasingly more likely to occur the more OBS scenes and sources that there are, which suggests some kind of race condition.

https://obsproject.com/logs/HIwvfa0gw0_Qnr5k

## Steps to Reproduce

1.  Create a new OBS scene collection "Cam test 1"
2.  In the default scene, add 2 or 3 video capture devices of different webcams.  I am using 2 C920 cameras here in my reproduction testing.
3.  Now add a media source and choose a MP4 video to play in a loop, and enable hardware decoding.  (This isn't necessarily relevant to the problem but I had to add some scene items and this seemed to be just enough to trigger the problem)
4.  Resize both webcam sources and the media source so you can see all 3 of them successfully in the preview window.
5. Duplicate the scene collection as "Cam test 2".
6. Switch back and forth between "Cam test 1" and "Cam test 2", and observe how one or both webcams no longer work in one of the collections, but may still work in the other collection.
7. If the bug does not show itself, delete "Cam test 2" scene collection and switch back to the "Cam test 1" collection.  Now add an additional Media source of another MP4 file and resize it so all sources are simultaneously visible.
8. Go to step 5 again.  Repeat steps 5-8 a few times if necessary and the bug should show itself by causing one or both webcams to no longer work in one of the scene collections.

## Additional information

This appears to be a bug in which OBS is loading the new scene collection before it has been able to completely unload the first collection leading to a race condition of some sort in which the webcam device is still opened from the first collection causing trying to open it to fail when the second one is loaded.

Matt WizardCM suggested as a troubleshooting workaround to create a new scene collection named "Blank" which contains nothing, and switch to it first, then switch to the desired collection each time.  I tried this and it does seem to work around the race condition and allow both webcams to work.

In summary, switching:
From Cam test 1 to Cam test 2 -> results in one or both webcams failing to start up.
From Cam test 1 to Blank to Cam test 2 -> results in both cams working properly as it adds additional time for OBS to complete unloading the first collection and disconnect the webcam before trying to load it in the new collection.

**Comment by skeletonbow** [comment]
Adding link to a discussion I hadd with @WizardCM in Discord a while back about this issue in case it's helpful for developers looking into the issue or other users understanding the issue.  Potential temporary workarounds are suggested which while not ideal, did work around the issue.

https://discordapp.com/channels/348973006581923840/374635871808520195/694091488463945729

**Comment by skeletonbow** [comment]
Adding some potentially useful excerpts of the conversation on Discord with Matt upon RytoEX's request.

`[4:20 AM] Matt | WizardCM: @skeletonbow i can definitely repro the webcam issue
...
[5:07 AM] Matt | WizardCM: @skeletonbow everything occurs in one function, pretty much. it fetches the new scene collection's config, cleans up the UI, loops through all the sources and runs obs_source_remove(), then loads all sources via obs_load_sources(). There's no timed delay or anything to give time for sources to properly shut down before they're re-initialised
[5:08 AM] Matt | WizardCM: at the very least, I would say it should count the number of sources in the old collection, wait based on that number, then load the new sources
[5:12 AM] Matt | WizardCM: a real fix would be for sources to announce when they've fully finished unloading resources and then going based on that
`

**Comment by RBswcx** [comment]
I seem to be having trouble with this as well. I have two different web cameras, a C922 and a Nexigo. I use them as webcams in different scenes. When switching between scenes that have the different cameras I have to deactivate and then activate the camera in the scene to get it to start the camera. If I switch to a different scene that uses the same camera, that works fine. It's only when the destination scene is using the other camera that I have to switch it off and on. I have both cameras set to be deactivated when not being viewed. Maybe it's this issue or maybe there isn't enough transition delay to allow the running camera to deactivate before the new scene loads??

**Comment by jp9000** [comment]
Currently working on a potential solution for this due to a coinciding issue

**Comment by lindenkron** [comment]
Would this also be the fix for large Scene Collections crashing when switching from one large to another?

Currently a lot of people are using a 'clean' Scene Collection as an intermediate to avoid OBS indefinite hanging/crashing.

**Comment by jp9000** [comment]
Should be fixed with https://github.com/obsproject/obs-studio/commit/8b3416c1e70bd4dad6baf1d424357a6af540efe0
